crates/sui-framework/packages/dwallet-system/sources/dwallet.move

```diff
@@ -0,0 +1,238 @@
+        message_approval.message
+    }
+
+    public fun remove(message_approval: MessageApproval): (ID, vector<u8>) {
+        let MessageApproval {
+            dwallet_cap_id,
+            message
+        } = message_approval;
+        (dwallet_cap_id, message)
+    }
+
+    public fun create_shared_sign_messages<S: store, E: store + copy + drop>(sign_messages: SignMessages<S, E>, ctx: &mut TxContext) {
+        let holder = SharedSignMessages {
+            id: object::new(ctx),
+            sign_messages,
+        };
+        transfer::share_object(holder);
+    }
+
+    public fun remove_shared_sign_messages<S: store, E: store + copy + drop>(shared: SharedSignMessages<S, E>): SignMessages<S, E> {
+        let SharedSignMessages {
+            id,
+            sign_messages,
+        } = shared;
+        object::delete(id);
+        sign_messages
+    }
+
+    public(friend) fun create_sign_messages<S: store, E: store + copy + drop>(dwallet_id: ID, dwallet_cap_id: ID, messages: vector<vector<u8>>, sign_data: S, sign_data_event: E, ctx: &mut TxContext): SignMessages<S, E> {
+        SignMessages {
+            id: object::new(ctx),
+            dwallet_id,
+            dwallet_cap_id,
+            messages,
+            sign_data,
+            sign_data_event,
+        }
+    }
+
+    public fun sign_messages_dwallet_id<S: store, E: store + copy + drop>(sign_messages: &SignMessages<S, E>): ID {
+        sign_messages.dwallet_id
+    }
+
+    public fun sign_messages_dwallet_cap_id<S: store, E: store + copy + drop>(sign_messages: &SignMessages<S, E>): ID {
+        sign_messages.dwallet_cap_id
+    }
+
+    public fun sign_messages_messages<S: store, E: store + copy + drop>(sign_messages: &SignMessages<S, E>): vector<vector<u8>> {
+        sign_messages.messages
+    }
+
+    public fun sign_messages<S: store, E: store + copy + drop>(sign_messages: SignMessages<S, E>, message_approvals: vector<MessageApproval>, ctx: &mut TxContext) {
+        let SignMessages {
+            id,
+            dwallet_id,
+            dwallet_cap_id,
+            messages,
+            sign_data,
+            sign_data_event,
+        } = sign_messages;
+
+        object::delete(id);
+        let i = 0;
+        let messages_len = vector::length(&messages);
+        let approval_len = vector::length(&message_approvals);
+        assert!(messages_len == approval_len, EMesssageApprovalDWalletMismatch);
+
+        while (i < messages_len) {
+            let message_approval = vector::pop_back(&mut message_approvals);
+            let (message_approval_dwallet_cap_id, approved_message) = remove(message_approval);
+            assert!(dwallet_cap_id == message_approval_dwallet_cap_id, EMesssageApprovalDWalletMismatch);
+            let message = vector::borrow(&messages, i);
+            assert!(message == &approved_message, EMesssageApprovalDWalletMismatch);
+            i = i +1;
+        };
```

Comment:
Wouldn't it be more clean to have `create_sign_messages` get `message_approvals` so that we never even create a `SignMessages` object for un-approved messages, or am I missing something?